src/rslice.rs

```diff
@@ -0,0 +1,184 @@
+use libc::{c_char, size_t};
+use std::convert::{TryFrom, TryInto};
+use std::marker::PhantomData;
+
+/// A read-only view on a Rust byte slice.
+///
+/// This is used to pass data from crustls to callback functions provided
+/// by the user of the API.
+/// `len` indicates the number of bytes than can be safely read.
+///
+/// The memory exposed is available as specified by the function
+/// using this in its signature. For instance, when this is a parameter to a
+/// callback, the lifetime will usually be the duration of the callback.
+/// Functions that receive one of these must not retain copies of any of the
+/// pointers beyond the allowed lifetime.
+#[repr(C)]
+pub struct rustls_slice_bytes<'a> {
+    pub data: *const u8,
+    pub len: size_t,
+    phantom: PhantomData<&'a [u8]>,
+}
+
+impl<'a> From<&'a [u8]> for rustls_slice_bytes<'a> {
+    fn from(s: &[u8]) -> Self {
```

Comment:
My intuition is that you need `s` (which is a reference) to at least outlive `'a`. Is that lifetime bound implicit here?

Reply:
Yep, this is lifetime elision. https://doc.rust-lang.org/nomicon/lifetime-elision.html

> If there is exactly one input lifetime position (elided or not), that lifetime is assigned to all elided output lifetimes.

Reply:
Moreover, the lifetimes are defined to be equal in the impl line above (the lifetime of the slice that's the parameter to From, and the lifetime parameter of rustls_slice_bytes), and the compiler will check that the function signature here (after applying lifetime elision rules) matches the signature of From::from with the type parameters substituted in.